app/maptools/measurementmaptool.cpp

```diff
@@ -0,0 +1,274 @@
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "measurementmaptool.h"
+
+MeasurementMapTool::MeasurementMapTool( QObject *parent )
+  : AbstractMapTool{ parent }
+{
+  connect( this, &AbstractMapTool::mapSettingsChanged, this, &MeasurementMapTool::onMapSettingsChanged );
+
+  if ( mapSettings() )
+  {
+    onMapSettingsChanged( mapSettings() );
+  }
+}
+
+MeasurementMapTool::~MeasurementMapTool() = default;
+
+void MeasurementMapTool::addPoint()
+{
+  QgsPoint transformedPoint = mapSettings()->screenToCoordinate( mCrosshairPoint );
+  mPoints.push_back( transformedPoint );
+  rebuildGeometry();
+}
+
+void MeasurementMapTool::removePoint()
+{
+  if ( !mPoints.isEmpty() )
+  {
+    mPoints.pop_back();
+    checkCanCloseShape();
+    rebuildGeometry();
+  }
+}
+
+void MeasurementMapTool::updateDistance()
+{
+  if ( mPoints.isEmpty() )
+  {
+    setLengthWithGuideline( 0.0 );
+    return;
+  }
+
+  checkCanCloseShape();
+
+  QgsPoint lastPoint = mPoints.last();
+  QgsPoint transformedCrosshairPoint = mapSettings()->screenToCoordinate( mCrosshairPoint );
+
+  double calculatedLength = mDistanceArea.measureLength( mRecordedGeometry ) + mDistanceArea.measureLine( transformedCrosshairPoint, lastPoint );
+  setLengthWithGuideline( calculatedLength );
+}
+
+void MeasurementMapTool::checkCanCloseShape()
+{
+  if ( mRecordedGeometry.isEmpty() || mPoints.count() < 3 )
+  {
+    setCanCloseShape( false );
+    return;
+  }
+
+  if ( mPoints.count() >= 3 )
+  {
+    QgsPoint firstPoint = mPoints.first();
+    QPointF firstPointScreen = mapSettings()->coordinateToScreen( firstPoint );
+    double distanceToFirstPoint = std::hypot( mCrosshairPoint.x() - firstPointScreen.x(), mCrosshairPoint.y() - firstPointScreen.y() );
+    setCanCloseShape( distanceToFirstPoint <= CLOSE_THRESHOLD );
+  }
+}
+
+void MeasurementMapTool::closeShape()
+{
+  if ( mPoints.count() < 3 )
+    return;
+
+  QList<QgsPointXY> pointList;
+  for ( const QgsPoint &point : mPoints )
+    pointList.append( QgsPointXY( point.x(), point.y() ) );
+
+  QgsGeometry polygonGeometry = QgsGeometry::fromPolygonXY( QList<QList<QgsPointXY>>() << pointList );
+  setRecordedGeometry( polygonGeometry );
+
+  double area = mDistanceArea.measureArea( polygonGeometry );
+  setArea( area );
+
+  double perimeter = mDistanceArea.measurePerimeter( polygonGeometry );
+  setPerimeter( perimeter );
+
+  setCanCloseShape( false );
+  setCloseShapeDone( true );
+}
+
+void MeasurementMapTool::reset()
```

Comment:
I have a feeling I asked you previously to rename this function to `reset`, but looking at it now, I think it makes more sense to name it `resetMeasurement` (or something similar ofc). The thing is `reset` functions are commonly used to reset everything in the object to its initial state - including member pointers and so on. Here you only try to reset the abacus :) Sorry for the confusion!